### DISCUSSION TITLE
Allow adding recipients retroactively

### DISCUSSION THREAD

**Discussion by feefladder** [openpgpjs/openpgpjs#1604] [Ideas]
I'm working on a way to share files in an end-to-end encrypted way, possibly using PGP. The files are rather big and therefore re-encrypting the file would take pretty long. What I'm looking to implement is something like this:
![E2E_add_recipient](https://user-images.githubusercontent.com/33122845/220954960-589df94a-08bc-47bc-905a-e192b5cf7eee.png)
It is important that files on the server are PGP encrypted, because it is a standard, and I would like to make a module that allows for decryption using hardware keys like so:
![E2E_decrypt](https://user-images.githubusercontent.com/33122845/220958416-c38d1132-9746-451e-a9f1-aeb0636a769b.png)
Now again, because these are large files, it would cost the Data Provider quite some bandwidth to: 

1. Download the data
2. re-encrypt the data with Data User's public key
3. re-upload the data that is encrypted using a different AES key

According to [this answer](https://security.stackexchange.com/a/244951/289735) it is technically possible, but not supported by tools. I would be willing to help work on this if possible :)

**Comment by larabr** [comment]
I haven't looked at the schema in detail, so sorry if I have misunderstood the use case, but OpenPGP.js already supports working with session keys directly: you can generate/encrypt/decrypt them (via `generateSessionKey`/`encryptionSessionKey`/`decryptSessionKeys`), as well as using them to encrypt/decrypt message data.

For instance, as far as the requirement of reusing a decrypted session key goes, if you have an `armoredMessage` that includes encrypted session keys & data, I think it should work to:
1. Decrypt session key only: `const sessionKeys = await decryptSessionKeys({ message: await readMessage({ armoredMessage }), decryptionKeys: [...] })`;
2. Decrypt message data with retrieved session key(s): `const decryptionResult = await decrypt({ message: await readMessage({ armoredMessage }), sessionKeys })`
3. Re-encrypt the session key to different public keys: `const encryptSessionKeys = await encryptSessionKey({ ...sessionKeys[0], encryptionKeys })`

Attaching the `encryptedSessionKeys` to the original `message` is slightly more involved. If you have the option of storing encrypted session key data separately from encrypted message data, that would make the code a bit simpler.
Does this cover your use case?

Regarding hardware keys, we are currently working on an API : #1567 , you can check if the proposed interface would work for you 🙂
